### ISSUE TITLE
blocks_in_conditions gives wrong spans and suggestions

### ISSUE THREAD

**Issue by hellow554** [rust-lang/rust-clippy#12162]
### Summary

The code below gives a) a wrong warning (it should not be emitted), and b) an invalid suggestion.

I think both can be fixed by ignoring macro expansions in this case, what do you think?

### Lint Name

blocks_in_conditions

### Reproducer

I tried this code:

```rust
macro_rules! timed {
    ($name:expr, $body:expr $(,)?) => {{
        let __scope = ();
        $body
    }};
}

fn main() {
    if timed!("check this!", false) {
        println!();
    }
}

```

I saw this happen:

```
warning: in an `if` condition, avoid complex blocks or closures with blocks; instead, move the block or closure higher and bind it with a `let`
  --> src/main.rs:6:6
  |
5 |       }};
  |  ______^
6 | | }
7 | |
8 | | fn main() {
9 | |     if timed!("check this!", false) {
  | |____^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#blocks_in_conditions
  = note: `#[warn(clippy::blocks_in_conditions)]` on by default
help: try
   |
5  ~     }let res = {
6  +         let __scope = ();
7  +         $body
8  ~     }; if resif timed!("check this!", false) {
```

### Version

```text
rustc -Vv
rustc 1.77.0-nightly (098d4fd74 2024-01-16)
binary: rustc
commit-hash: 098d4fd74c078b12bfc2e9438a2a04bc18b393bc
commit-date: 2024-01-16
host: x86_64-unknown-linux-gnu
release: 1.77.0-nightly
LLVM version: 17.0.6
```

### Additional Labels

@rustbot label I-suggestion-causes-error
